exp/tests/test_response_views.py

```diff
@@ -846,8 +949,254 @@ def test_get_exit_survey_fields_in_summary_json(self):
+        query_string = urlencode(
+            {"demo_options": ["participant__global_id"]}, doseq=True
+        )
+        response = self.client.get(f"{demographic_json_url}?{query_string}")
+        content = response.content.decode("utf-8")
+        data = json.loads(content)
+        for demo in data:
+            self.assertIn("global_id", demo["participant"])
+            self.assertEqual(
+                demo["participant"]["global_id"],
+                str(
+                    Response.objects.get(uuid=demo["response"]["uuid"]).child.user.uuid
+                ),
+            )
+
+        # Without participant__global_id selected, this info is absent
+        query_string = urlencode({"demo_options": []}, doseq=True)
+        response = self.client.get(f"{demographic_json_url}?{query_string}")
+        content = response.content.decode("utf-8")
+        data = json.loads(content)
+        for demo in data:
+            self.assertNotIn("global_id", demo["participant"])
+            self.assertNotIn(
+                str(
+                    Response.objects.get(uuid=demo["response"]["uuid"]).child.user.uuid
+                ),
+                content,
+            )
+
+    def test_get_appropriate_individual_responses_as_researcher(self):
+        self.client.force_login(self.study_reader)
+        response = self.client.get(
+            reverse("exp:study-responses-list", kwargs={"pk": self.study.pk})
+        )
+        content = response.content.decode("utf-8")
+
+        matches = re.finditer('data-response-uuid="(.*)"', content)
+        n_matches = 0
+        for m in matches:
+            n_matches += 1
+            this_response_uuid = m.group(1)
+            response = Response.objects.get(uuid=this_response_uuid)
+            self.assertEqual(response.study.pk, self.study.pk)
+            self.assertTrue(response.has_valid_consent)
+        self.assertTrue(
+            n_matches == self.n_responses + self.n_previews
+            or (n_matches == 10 and (self.n_responses + self.n_previews) > 10)
+        )
+
+        self.assertNotIn(self.poison_string, content)
+
+    def test_get_appropriate_individual_responses_as_previewer(self):
+        self.client.force_login(self.study_previewer)
+        response = self.client.get(
+            reverse("exp:study-responses-list", kwargs={"pk": self.study.pk})
+        )
+        content = response.content.decode("utf-8")
+
+        matches = re.finditer('data-response-uuid="(.*)"', content)
+        n_matches = 0
+        for m in matches:
+            n_matches += 1
+            this_response_uuid = m.group(1)
+            response = Response.objects.get(uuid=this_response_uuid)
+            self.assertEqual(response.study.pk, self.study.pk)
+            self.assertTrue(response.has_valid_consent)
+            self.assertTrue(response.is_preview)
+        self.assertTrue(
+            n_matches == self.n_previews or (n_matches == 10 and self.n_previews > 10)
```

Comment:
Why is there non-determinism in this test?

Reply:
This isn't really non-deterministic (result won't vary with same input), just doesn't break if we were to add more responses during setup in the future. But agreed it makes more sense just to adjust tests in that case so we don't silently stop checking for exact equality - fixed.